Android/AppsFlyerURLResolver/AppsFlyerURLResolverLib/src/main/java/com/appsflyer/resolver/URLResolver.kt

```diff
@@ -1,66 +1,96 @@
 
     init {
         CookieHandler.setDefault(CookieManager())
     }
 
     fun resolve(url: String?, maxRedirections: Int = 10, urlResolverListener: URLResolverListener) {
-        Thread {
-            if (url == null) {
-                urlResolverListener.onComplete(null)
-                return@Thread
-            }
-            afDebugLog("resolving $url")
-            if (url.isValidURL()) {
-                val redirects = ArrayList<String>().apply {
-                    add(url)
-                }
+        startResolvingJob(url, urlResolverListener) {
+            it?.run {
+                var redirect = this
                 var res: AFHttpResponse? = null
                 for (i in 0 until maxRedirections) {
                     // resolve current URL - check for redirection
-                    res = resolveInternal(redirects.last())
-                    res.redirected?.let { // if redirected to another URL
-                        redirects.add(it)
-                    } ?: break
+                    res = resolveInternal(redirect) { con ->
+                        if (con.isRedirected) {
+                            // redirect
+                            con.redirectedTo.also { l ->
+                                afDebugLog("redirecting to $l")
+                            }
+                        } else {
+                            // did not redirected
+                            null
+                        }
+                    }
+                    res.redirected?.let { r ->
+                        redirect = r // if redirected to another URL - update last url
+                    } ?: break // break will not work inside `also` block on `resolveInternal`
                 }
-                if (res?.error == null) {
-                    afDebugLog("found link: ${redirects.last()}")
-                    urlResolverListener.onComplete(redirects.last())
-                } else {
-                    urlResolverListener.onComplete(null)
+                res?.apply {
+                    // return the last link we found
+                    redirected = redirect
+                }
+            }
+        }
+    }
+
+    fun resolveJSRedirection(url: String?, urlResolverListener: URLResolverListener) {
+        startResolvingJob(url, urlResolverListener) {
+            it?.run {
+                resolveInternal(this) { con ->
+                    if (con.isSuccessful && con.isHtmlPage) {
+                        con.jsRedirected
+                    } else {
+                        null
+                    }
                 }
-            } else {
-                urlResolverListener.onComplete(url)
             }
-        }.start()
+        }
     }
 
-    private fun resolveInternal(uri: String): AFHttpResponse {
+    private fun startResolvingJob(
+        url: String?,
+        urlResolverListener: URLResolverListener,
+        job: ResolveJob
+    ) {
+        Executors.newSingleThreadExecutor().submit {
```

Comment:
please move this executor to the class property instead of creating new one for each resolve. I believe `newCachedThreadPool()` should do better job here.

Reply:
done